colossalai/nn/optimizer/distributed_lamb.py

```diff
@@ -0,0 +1,167 @@
+# Disclaimer: Modified from https://github.com/NUS-HPC-AI-Lab/pytorch-lamb/blob/master/optim/lamb.py
+
+
+from typing import Dict
+
+import torch
+import torch.distributed as dist
+from torch.optim import Optimizer
+
+from colossalai.tensor.d_tensor import is_distributed_tensor
+
+__all__ = ["DistributedLamb"]
+
+
+class DistributedLamb(Optimizer):
+    r"""Implements the Lamb algorithm, with extra support for ZeRO 2 and Tensor Parallel.
+    Proposed in `Large Batch Optimization for Deep Learning: Training BERT in 76 minutes`_.
+    Example (4 devices):
+        >>> device_mesh, tp_group, dp_group = DistributedLamb.set_distributed(2, 2)
+        >>> optim = DistributedLamb(model.parameters(), lr=1e-3, device_mesh=device_mesh)
+
+    Arguments:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float, optional): learning rate (default: 1e-3)
+        betas (Tuple[float, float], optional): coefficients used for computing
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+    .. _Large Batch Optimization for Deep Learning: Training BERT in 76 minutes:
+        https://arxiv.org/abs/1904.00962
+    """
+
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-6,
+        weight_decay=0,
+        bias_correction=True,
+    ):
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))
+
+        # self.setup_distributed(tp_group, dp_group)
+        self.shard_to_param = {}
+        self.tp_size = self.dp_size = 1
+        defaults = dict(lr=lr, betas=betas, eps=eps, weight_decay=weight_decay, bias_correction=bias_correction)
+        super().__init__(params, defaults)
+
+    def setup_distributed(
+        self, tp_group: dist.ProcessGroup = None, dp_group: dist.ProcessGroup = None, shard_to_param: Dict = None
```

Comment:
Wrong type annotations. They should be `Optional[dist.ProcessGroup]` and `Optional[Dict]`

Reply:
😅